nilearn/signal.py

```diff
@@ -833,14 +845,30 @@ def _censor_signals(signals, confounds, sample_mask):
     return signals, confounds
 
 
-def _interpolate_volumes(volumes, sample_mask, t_r):
+def _interpolate_volumes(volumes, sample_mask, t_r, extrapolate):
     """Interpolate censored volumes in signals/confounds."""
+    if extrapolate:
+        extrapolate_default = (
+            "By default the cubic spline interpolator extrapolates "
+            "the out-of-bounds censored volumes in the data run. This "
+            "can lead to undesired filtered signal results. In future "
+            "releases, the default strategy will be not to extrapolate "
+            "and discard those volumes at filtering."
+        )
+        warnings.warn(
+            category=FutureWarning,
+            message=extrapolate_default,
+        )
     frame_times = np.arange(volumes.shape[0]) * t_r
     remained_vol = frame_times[sample_mask]
     remained_x = volumes[sample_mask, :]
-    cubic_spline_fitter = CubicSpline(remained_vol, remained_x)
+    cubic_spline_fitter = CubicSpline(
+        remained_vol, remained_x, extrapolate=extrapolate
+    )
     volumes_interpolated = cubic_spline_fitter(frame_times)
     volumes[~sample_mask, :] = volumes_interpolated[~sample_mask, :]
+    # discard non-interpolated out-of-bounds volumes
+    volumes = volumes[~np.isnan(volumes).all(axis=1), :]
```

Comment:
> What I recommand is to discard the NaN volumes outside of `_interpolate_volumes`, in `_handle_scrubbed_volumes`. And since sample_mask will be modified, `_handle_scrubbed_volumes` will return signals, confounds, sample_mask. 

Done, see https://github.com/nilearn/nilearn/pull/4028/commits/2ae9efdd0b0d5393566aed23ce430eeb403ad8c0.